File: 0085-maximal-rectangle/0085-maximal-rectangle.cs
public class Solution {
    public int MaximalRectangle(char[][] matrix) {
        List<List<int>> dp = new List<List<int>>();
        int l = matrix[0].Length;
        int i = 0;
        int j = 0;
        int res = 0;
        
        foreach(char[] m in matrix){
            foreach(List<int>r in dp){
                int start = r[0];
                int end = r[1];
                
                if(start != -1){
                     for(int k = start; k <= end; k++){
                        if(m[k] != '1'){
                            if(k == start){
                                r[0] = -1;
                            }else{
                                r[2] ++;
                                r[1] = k - 1;
                            }
                            break;
                        }

                        if(k == end){
                            r[2] ++;
                        }
                    }
                }
                
            }
            
            i = 0;
            j = 0;
            
            while(i < l){
                if(m[i] == '0'){
                    i++;
                }else{
                    j = i;
                    while(j < l){
                        j++;
                        
                        if(j == l || m[j] == '0'){
                            dp.Add(new List<int>{i, j - 1, 1});
                            break;
                        }
                    }
                    i++;
                }
            }
            
            foreach(List<int> r in dp){
                if(r[0] != -1){
                    res = Math.Max(res, r[2] * (r[1] - r[0] + 1));
                }
                
                // Console.WriteLine(r[0]);
                // Console.WriteLine(r[1]);
                // Console.WriteLine(r[2]);
                // Console.WriteLine();
                
            }
//             Console.WriteLine("***");
            
            
//             Console.WriteLine(res);
            
//             Console.WriteLine("---------");

        }
        
        
        
        return res;
    }
}

File: 0085-maximal-rectangle/0085-maximal-rectangle.java
class Solution {
    public int maximalRectangle(char[][] matrix) {
       if (matrix.length == 0 || matrix[0].length == 0)
               return 0;
       var dp = new int[matrix[0].length];
       var maxArea = 0;

       for (var row : matrix) {
               for (var j = 0; j < matrix[0].length; j++)
                       dp[j] = (row[j] == '1') ? dp[j] + 1 : 0;
               maxArea = Math.max(maxArea, largestRectangleArea(dp));
       }
       return maxArea;
}

// This method is straight copied from LC 84
private int largestRectangleArea(int[] heights) {
       var n = heights.length;
       var leftSmaller = new int[n];
       var rightSmaller = new int[n];
       leftSmaller[0] = -1;
       rightSmaller[n - 1] = n;

       for (var i = 1; i < n; i++) {
   		var j = i - 1;
   		while (j >= 0 && heights[j] >= heights[i])
   			j = leftSmaller[j];
   		leftSmaller[i] = j;
       }

       for (var i = n - 2; i >= 0; i--) {
   		var j = i + 1;
   		while (j < n && heights[j] >= heights[i])
   			j = rightSmaller[j];
   		rightSmaller[i] = j;
       }

       var maxArea = 0;
       for (var i = 0; i < n; i++)
   		maxArea = Math.max(maxArea, heights[i] * (rightSmaller[i] - leftSmaller[i] - 1));
       return maxArea;
}
}